File: app/src/main/java/ru/michanic/mymot/Interactors/ApiInteractor.java
package ru.michanic.mymot.Interactors;

import android.util.Log;

import java.util.List;

import io.realm.Realm;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import ru.michanic.mymot.Kotlin.Models.AppPageText;
import ru.michanic.mymot.Kotlin.Models.Category;
import ru.michanic.mymot.Kotlin.Models.Location;
import ru.michanic.mymot.Kotlin.Models.Manufacturer;
import ru.michanic.mymot.Kotlin.Models.Model;
import ru.michanic.mymot.Kotlin.Models.ModelDetails;
import ru.michanic.mymot.Kotlin.Models.Volume;
import ru.michanic.mymot.MyMotApplication;
import ru.michanic.mymot.Protocols.ApiInterface;
import ru.michanic.mymot.Protocols.Const;
import ru.michanic.mymot.Protocols.LoadingInterface;
import ru.michanic.mymot.Protocols.LoadingModelDetailsInterface;
import ru.michanic.mymot.Protocols.LoadingTextInterface;
import ru.michanic.mymot.Utils.DataManager;

public class ApiInteractor {

    //private Gson gson = new GsonBuilder().create();
    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(Const.API_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    private ApiInterface apiInterface = retrofit.create(ApiInterface.class);
    private Realm realm;
    private DataManager dataManager = new DataManager();

    public ApiInteractor() {
        Realm.init(MyMotApplication.appContext);
        realm = Realm.getDefaultInstance();
    }

    public void loadData(final LoadingInterface loadingInterface) {

        dataManager.cleanAdverts();
        //MyMotApplication.configStorage.clearCsrfToken();

        loadExteptedWords(new LoadingInterface() {

            @Override
            public void onLoaded() {
                loadAboutText(new LoadingInterface() {

                    @Override
                    public void onLoaded() {
                        loadRegions(new LoadingInterface() {

                            @Override
                            public void onLoaded() {
                                loadVolumes(new LoadingInterface() {

                                    @Override
                                    public void onLoaded() {
                                        loadClasses(new LoadingInterface() {

                                            @Override
                                            public void onLoaded() {
                                                loadModels(new LoadingInterface() {

                                                    @Override
                                                    public void onLoaded() {
                                                        dataManager.assignCategories();
                                                        loadingInterface.onLoaded();
                                                    }
                                                    @Override
                                                    public void onFailed() {
                                                        loadingInterface.onFailed();
                                                    }
                                                });

                                            }
                                            @Override
                                            public void onFailed() {
                                                loadingInterface.onFailed();
                                            }

                                        });

                                    }

                                    @Override
                                    public void onFailed() {
                                        loadingInterface.onFailed();
                                    }
                                });

                            }

                            @Override
                            public void onFailed() {
                                loadingInterface.onFailed();
                            }

                        });
                    }
                    @Override
                    public void onFailed() {
                        loadingInterface.onFailed();
                    }

                });
            }
            @Override
            public void onFailed() {
                loadingInterface.onFailed();
            }
        });

    }

    private void loadExteptedWords(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadExteptedWords");

        apiInterface.loadExteptedWords().enqueue(new Callback<List<String>>() {
            @Override
            public void onResponse(Call<List<String>> call, Response<List<String>> response) {
                MyMotApplication.getConfigStorage().exteptedWords = response.body();
                loadingInterface.onLoaded();
                Log.e("loadData", "words loaded");
            }

            @Override
            public void onFailure(Call<List<String>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("loadData", t.toString());
            }
        });

    }

    private void loadAboutText(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadAboutText");

        apiInterface.loadAboutText().enqueue(new Callback<AppPageText>() {
            @Override
            public void onResponse(Call<AppPageText> call, Response<AppPageText> response) {
                MyMotApplication.getConfigStorage().aboutText = response.body().getText();
                loadingInterface.onLoaded();
                Log.e("loadData", "about loaded");
            }

            @Override
            public void onFailure(Call<AppPageText> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });

    }


    private void loadRegions(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadRegions");

        apiInterface.loadRegions().enqueue(new Callback<List<Location>>() {
            @Override
            public void onResponse(Call<List<Location>> call, Response<List<Location>> response) {

                realm.beginTransaction();
                realm.copyToRealmOrUpdate(response.body());
                realm.commitTransaction();

                loadingInterface.onLoaded();
                Log.e("loadData", "regions loaded");
            }

            @Override
            public void onFailure(Call<List<Location>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });
    }

    public void loadRegionCities(final Location region, final LoadingInterface loadingInterface) {

        apiInterface.loadRegionCities(region.getId()).enqueue(new Callback<List<Location>>() {
            @Override
            public void onResponse(Call<List<Location>> call, Response<List<Location>> response) {
                realm.beginTransaction();
                realm.copyToRealmOrUpdate(response.body());
                realm.commitTransaction();
                loadingInterface.onLoaded();
            }

            @Override
            public void onFailure(Call<List<Location>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });
    }

    private void loadVolumes(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadVolumes");

        apiInterface.loadVolumes().enqueue(new Callback<List<Volume>>() {
            @Override
            public void onResponse(Call<List<Volume>> call, Response<List<Volume>> response) {

                realm.beginTransaction();
                realm.copyToRealmOrUpdate(response.body());
                realm.commitTransaction();

                loadingInterface.onLoaded();
                Log.e("loadData", "volumes loaded");
            }

            @Override
            public void onFailure(Call<List<Volume>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });
    }

    private void loadClasses(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadClasses");

        apiInterface.loadClasses().enqueue(new Callback<List<Category>>() {
            @Override
            public void onResponse(Call<List<Category>> call, Response<List<Category>> response) {

                realm.beginTransaction();
                realm.copyToRealmOrUpdate(response.body());
                realm.commitTransaction();

                loadingInterface.onLoaded();
                Log.e("loadData", "classes loaded");
            }

            @Override
            public void onFailure(Call<List<Category>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });

    }

    private void loadModels(final LoadingInterface loadingInterface) {
        Log.e("loadData", "loadModels");

        apiInterface.loadModels().enqueue(new Callback<List<Manufacturer>>() {
            @Override
            public void onResponse(Call<List<Manufacturer>> call, Response<List<Manufacturer>> response) {

                List<Integer> favoriteModels = dataManager.getFavouriteModelIDs();
                List<Volume> volumes = dataManager.getVolumes();

                realm.beginTransaction();
                for (Manufacturer manufacturer: response.body()) {

                    Log.e("manufacturer", manufacturer.getName());

                    for (Model model: manufacturer.getModels()) {

                        String volumeText = model.getVolume();
                        float volumeVal = 0;
                        try {
                            volumeVal = Float.parseFloat(volumeText.replace("от ", ""));
                        } catch (NumberFormatException nfe) { System.out.println("NumberFormatException: " + nfe.getMessage()); }
                        model.setVolume_value(volumeVal);
                        model.setVolume(volumeText + " куб.см.");

                        for (Volume volume: volumes) {
                            if ((float)volume.getMin() <= volumeVal && volumeVal <= (float)volume.getMax()) {
                                model.setVolume_id(volume.getId());
                                break;
                            }
                        }

                        model.setFavourite(favoriteModels.contains(model.getId()));
                        model.setManufacturer(manufacturer);

                        realm.copyToRealmOrUpdate(model);
                    }
                }
                realm.commitTransaction();

                loadingInterface.onLoaded();
                Log.e("loadData", "models loaded");
            }

            @Override
            public void onFailure(Call<List<Manufacturer>> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });

    }

    public void loadModelDetails(int modelId, final LoadingModelDetailsInterface loadingInterface) {
        Log.e("loadData", "loadModelDetails");

        apiInterface.loadModelDetails(modelId).enqueue(new Callback<ModelDetails>() {
            @Override
            public void onResponse(Call<ModelDetails> call, Response<ModelDetails> response) {
                loadingInterface.onLoaded(response.body());
            }

            @Override
            public void onFailure(Call<ModelDetails> call, Throwable t) {
                loadingInterface.onFailed();
                Log.e("response", t.toString());
            }
        });
    }

    public void loadAgreementText(final LoadingTextInterface loadingTextInterface) {
        apiInterface.loadAgreementText().enqueue(new Callback<AppPageText>() {
            @Override
            public void onResponse(Call<AppPageText> call, Response<AppPageText> response) {
                loadingTextInterface.onLoaded(response.body().getText());
            }

            @Override
            public void onFailure(Call<AppPageText> call, Throwable t) {
                loadingTextInterface.onLoaded(t.toString());
            }
        });
    }




}

File: app/src/main/java/ru/michanic/mymot/Kotlin/Interactors/ApiInteractor.kt
package ru.michanic.mymot.Kotlin.Interactors

import android.util.Log
import io.realm.Realm
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import ru.michanic.mymot.Kotlin.Models.*
import ru.michanic.mymot.Kotlin.MyMotApplication
import ru.michanic.mymot.Kotlin.Protocols.ApiInterface
import ru.michanic.mymot.Kotlin.Protocols.Const
import ru.michanic.mymot.Kotlin.Utils.ConfigStorage
import ru.michanic.mymot.Kotlin.Utils.DataManager
import kotlin.math.max
import kotlin.math.min

class ApiInteractor {

    private val retrofit = Retrofit.Builder()
        .baseUrl(Const.API_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
    private val apiInterface = retrofit.create(ApiInterface::class.java)
    private val realm: Realm
    private val dataManager = DataManager()
    private val configStorage = MyMotApplication.configStorage

    init {
        Realm.init(MyMotApplication.appContext)
        realm = Realm.getDefaultInstance()
    }

    fun loadData(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        dataManager.cleanAdverts()
        loadPropertyEnums({
            loadExteptedWords({
                loadAboutText({
                    loadRegions({
                        loadVolumes({
                            loadClasses({
                                loadModels({
                                    dataManager.assignCategories()
                                    onSuccess()
                                }, {
                                    onFail()
                                })
                            }, {
                                onFail()
                            })
                        }, {
                            onFail()
                        })
                    }, {
                        onFail()
                    })
                }, {
                    onFail()
                })
            }, {
                onFail()
            })
        }, {
            onFail()
        })

    }


    private fun loadExteptedWords(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadExteptedWords")
        apiInterface.loadExteptedWords()?.enqueue(object : Callback<List<String?>?> {
            override fun onResponse(
                call: Call<List<String?>?>,
                response: Response<List<String?>?>,
            ) {
                configStorage?.exteptedWords =
                    response.body()?.filterNotNull() ?: emptyList()
                onSuccess()
                Log.e("loadData", "words loaded")
            }

            override fun onFailure(call: Call<List<String?>?>, t: Throwable) {
                onFail()
                Log.e("loadData", t.toString())
            }
        })
    }

    private fun loadAboutText(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadAboutText")
        apiInterface.loadAboutText()?.enqueue(object : Callback<AppPageText?> {
            override fun onResponse(
                call: Call<AppPageText?>,
                response: Response<AppPageText?>
            ) {
                configStorage?.aboutText = response.body()?.text ?: ""
                onSuccess()
                Log.e("loadData", "about loaded")
            }

            override fun onFailure(call: Call<AppPageText?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }


    private fun loadRegions(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadRegions")
        apiInterface.loadRegions()?.enqueue(object : Callback<List<Location?>?> {
            override fun onResponse(
                call: Call<List<Location?>?>,
                response: Response<List<Location?>?>,
            ) {
                realm.beginTransaction()
                realm.copyToRealmOrUpdate(response.body())
                realm.commitTransaction()
                onSuccess()
                Log.e("loadData", "regions loaded")
            }

            override fun onFailure(call: Call<List<Location?>?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }

    fun loadRegionCities(region: Location, onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        apiInterface.loadRegionCities(region.id)
            ?.enqueue(object : Callback<List<Location?>?> {
                override fun onResponse(
                    call: Call<List<Location?>?>,
                    response: Response<List<Location?>?>,
                ) {
                    realm.beginTransaction()
                    realm.copyToRealmOrUpdate(response.body())
                    realm.commitTransaction()
                    onSuccess()
                }

                override fun onFailure(call: Call<List<Location?>?>, t: Throwable) {
                    onFail()
                    Log.e("response", t.toString())
                }
            })
    }

    private fun loadVolumes(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadVolumes")
        apiInterface.loadVolumes()?.enqueue(object : Callback<List<Volume?>?> {
            override fun onResponse(
                call: Call<List<Volume?>?>,
                response: Response<List<Volume?>?>,
            ) {
                realm.beginTransaction()
                realm.copyToRealmOrUpdate(response.body())
                realm.commitTransaction()
                onSuccess()
                Log.e("loadData", "volumes loaded")
            }

            override fun onFailure(call: Call<List<Volume?>?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }

    private fun loadClasses(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadClasses")
        apiInterface.loadClasses()?.enqueue(object : Callback<List<Category?>?> {
            override fun onResponse(
                call: Call<List<Category?>?>,
                response: Response<List<Category?>?>,
            ) {
                realm.beginTransaction()
                realm.copyToRealmOrUpdate(response.body())
                realm.commitTransaction()
                onSuccess()
                Log.e("loadData", "classes loaded")
            }

            override fun onFailure(call: Call<List<Category?>?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }

    private fun loadModels(onSuccess: (() -> Unit), onFail: (() -> Unit)) {
        Log.e("loadData", "loadModels")
        apiInterface.loadModels()?.enqueue(object : Callback<List<Manufacturer?>?> {
            override fun onResponse(
                call: Call<List<Manufacturer?>?>,
                response: Response<List<Manufacturer?>?>,
            ) {
                val favoriteModels = dataManager.favouriteModelIDs
                val volumes = dataManager.volumes
                var minPower: Double = 1000.0
                var maxPower: Double = 0.0
                realm.beginTransaction()
                for (manufacturer in response.body() ?: emptyList()) {
                    manufacturer?.name?.let { Log.e("manufacturer", it) }
                    for (model in manufacturer?.models ?: continue) {
                        val volumeText = model.volume
                        var volumeVal = 0f
                        try {
                            if (volumeText != null) {
                                volumeVal = volumeText.replace("от ", "").toFloat()
                            }
                        } catch (nfe: NumberFormatException) {
                            println("NumberFormatException: " + nfe.message)
                        }
                        model.setVolume_value(volumeVal)
                        model.volume = "$volumeText куб.см."
                        for (volume in volumes) {
                            if (volume.min.toFloat() <= volumeVal && volumeVal <= volume.max.toFloat()) {
                                model.volume_id = volume.id
                                break
                            }
                        }
                        model.isFavourite = favoriteModels.contains(model.id)
                        model.manufacturer = manufacturer
                        val powerRange = model.minMaxPower()
                        //println(model.name + " " + powerRange.first + "/" + powerRange.second + " current: " + minPower + "/" + maxPower)
                        if (powerRange.first != null) {
                            minPower = min(minPower, powerRange.first!!)
                        }
                        if (powerRange.second != null) {
                            maxPower = max(maxPower, powerRange.second!!)
                        }

                        realm.copyToRealmOrUpdate(model)
                    }
                }
                realm.commitTransaction()
                configStorage?.minPower = minPower
                configStorage?.maxPower = maxPower
                onSuccess()
                Log.e("loadData", "models loaded")
                Log.e("loadData", "minPower: " + minPower + " maxPower: " + maxPower)
            }

            override fun onFailure(call: Call<List<Manufacturer?>?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }

    fun loadModelDetails(
        modelId: Int,
        onSuccess: ((ModelDetails?) -> Unit),
        onFail: (() -> Unit)
    ) {
        apiInterface.loadModelDetails(modelId)?.enqueue(object : Callback<ModelDetails?> {
            override fun onResponse(
                call: Call<ModelDetails?>,
                response: Response<ModelDetails?>
            ) {
                onSuccess(response.body())
            }

            override fun onFailure(call: Call<ModelDetails?>, t: Throwable) {
                onFail()
                Log.e("response", t.toString())
            }
        })
    }

    fun loadPropertyEnums(
        onSuccess: (() -> Unit),
        onFail: (() -> Unit)
    ) {
        apiInterface.loadPropertyEnums().enqueue(object : Callback<PropertyEnums?> {
            override fun onResponse(call: Call<PropertyEnums?>, response: Response<PropertyEnums?>) {
                val emptyMap = mapOf<Int, String>()
                val placements = response.body()?.placements?.map { e -> e.key.toInt() to e.value }?.toMap()
                configStorage?.placements = placements ?: emptyMap

                val coolingTypes = response.body()?.cooling_types?.map { e -> e.key.toInt() to e.value }?.toMap()
                configStorage?.coolingTypes = coolingTypes ?: emptyMap

                val driveTypes = response.body()?.drive_types?.map { e -> e.key.toInt() to e.value }?.toMap()
                configStorage?.driveTypes = driveTypes ?: emptyMap
                onSuccess()
            }

            override fun onFailure(call: Call<PropertyEnums?>, t: Throwable) {
                onFail()
            }
        })
    }

    fun loadAgreementText(result: ((String?) -> Unit)) {
        apiInterface.loadAgreementText()?.enqueue(object : Callback<AppPageText?> {
            override fun onResponse(
                call: Call<AppPageText?>,
                response: Response<AppPageText?>
            ) {
                result(response.body()?.text)
            }

            override fun onFailure(call: Call<AppPageText?>, t: Throwable) {
                result(null)
            }
        })
    }
}